samples/net/mqtt/src/modules/network/network_wifi.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/zbus/zbus.h>
+#include <zephyr/net/net_if.h>
+#include <zephyr/net/wifi_mgmt.h>
+#include <zephyr/net/net_event.h>
+
+#include "message_channel.h"
+
+/* Register log module */
+LOG_MODULE_REGISTER(network, CONFIG_MQTT_SAMPLE_NETWORK_LOG_LEVEL);
+
+/* This module does not subscribe to any channels */
+
+BUILD_ASSERT((sizeof(CONFIG_MQTT_SAMPLE_NETWORK_WIFI_SSID) > 1), "SSID must be set");
+BUILD_ASSERT((sizeof(CONFIG_MQTT_SAMPLE_NETWORK_WIFI_PSK) > 1), "PSK must be set");
+
+#define MGMT_EVENTS (NET_EVENT_WIFI_SCAN_RESULT		| \
+		     NET_EVENT_WIFI_SCAN_DONE		| \
+		     NET_EVENT_WIFI_CONNECT_RESULT	| \
+		     NET_EVENT_WIFI_DISCONNECT_RESULT)
+
+static bool ssid_located;
+static struct net_mgmt_event_callback wifi_shell_mgmt_cb;
+static uint32_t scan_result;
+extern char *net_sprint_ll_addr_buf(const uint8_t *ll, uint8_t ll_len, char *buf, int buflen);
+
+static void scan(void)
+{
+	struct net_if *iface = net_if_get_default();
+
+	__ASSERT(iface, "Returned network interface is NULL");
+
+	LOG_INF("Scanning for Wi-Fi networks...");
+
+	ssid_located = false;
+
+	if (net_mgmt(NET_REQUEST_WIFI_SCAN, iface, NULL, 0)) {
+		LOG_ERR("Wi-Fi scan failed");
+		return;
+	}
+}
+
+static void connect(void)
+{
+	if (ssid_located) {
+		LOG_INF("SSID found during scan, proceeding to connect");
+	} else {
+		LOG_INF("SSID not found during scan, aborting connect");
+		return;
+	}
+
+	struct net_if *iface = net_if_get_default();
+
+	__ASSERT(iface, "Returned network interface is NULL");
+
+	LOG_INF("Connecting to SSID: %s", CONFIG_MQTT_SAMPLE_NETWORK_WIFI_SSID);
+
+	static struct wifi_connect_req_params cnx_params = {
+		.ssid = CONFIG_MQTT_SAMPLE_NETWORK_WIFI_SSID,
+		.ssid_length = sizeof(CONFIG_MQTT_SAMPLE_NETWORK_WIFI_SSID) - 1,
+		.psk = CONFIG_MQTT_SAMPLE_NETWORK_WIFI_PSK,
+		.psk_length = sizeof(CONFIG_MQTT_SAMPLE_NETWORK_WIFI_PSK) - 1,
+		.channel = WIFI_CHANNEL_ANY,
+		.security = WIFI_SECURITY_TYPE_PSK,
+		.timeout = CONFIG_MQTT_SAMPLE_NETWORK_WIFI_CONNECTION_REQUEST_TIMEOUT_SECONDS,
+	};
+
+	if (net_mgmt(NET_REQUEST_WIFI_CONNECT, iface,
```

Comment:
I'm pretty sure you don't have to manually scan before requesting connection. `wpa_supp` will automatically scan, connect, possibly reconnect for you.

Reply:
Yes, but I like the formatted overview of all the available networks. Has a nice effect.
Perhaps this is possible to output from `wpa_supp`?

Reply:
Fair enough! The networks are filtered at a pretty early stage when requesting connection. I don't think you can get a network overview without a manual scan.